tiledb/api/proxy/proxy.h

```diff
@@ -0,0 +1,917 @@
+      // Assert: event state is `Change`
+
+      auto new_state{transition_table[to_index(state_)][to_index(event)]};
+      auto action{action_table[to_index(state_)][to_index(event)]};
+      /*
+       * Actions each happen within a try-block to ensure state consistency in
+       * case of an exception.
+       */
+      switch (action) {
+        case ProxyAction::none:
+          break;
+        case ProxyAction::construct:
+          try {
+            if (do_construct()) {
+              // Use the ordinary state transition
+            } else {
+              // Construction did not succeed. Stay in the `nascent` state.
+              new_state = ProxyState::nascent;
+            }
+          } catch (...) {
+            new_state = ProxyState::aborted;
+            throw;
+          }
+          break;
+        case ProxyAction::access:
+          ++n_accessors;
+          break;
+        case ProxyAction::release:
+          --n_accessors;
+          if (n_accessors > 0) {
+            // Stay in the `access` state instead of moving back to `present`.
+            new_state = ProxyState::access;
+          }
+          break;
+        case ProxyAction::r_and_d:
+          --n_accessors;
+          if (n_accessors > 0) {
+            // Stay in the `last_access` state instead of moving back to
+            // `present`.
+            new_state = ProxyState::last_access;
+            break;
+          }
+          // Assert: n_accessors == 0
+          [[fallthrough]];
+        case ProxyAction::destroy:
+          try {
+            do_destroy();
+          } catch (...) {
+            new_state = ProxyState::error;
+            throw;
+          }
+          break;
+        case ProxyAction::shutdown:
+          try {
+            do_shutdown();
+          } catch (...) {
+            new_state = ProxyState::error;
+            throw;
+          }
+          break;
+        default:
+          throw std::logic_error("Not reachable");
+      }
+      /*
+       * Assign new state only after action returns without throwing.
+       */
+      state_ = new_state;
+
+      /*
+       * At this point we can return if the message queue is empty. We cannot,
+       * however, check the message queue reliably without a queue lock. And
+       * we can't acquire a queue lock from the `Changing` state because of the
+       * anti-deadlock ordering of the mutexes. Thus we have to unlock to check
+       * the queue and then re-lock the state mutex.
+       *
+       * Thus we cycle through event states until we're back at `Hand-off`.
+       * That's the required state at the top of the event loop. That state will
+       * also transition to `Change` as required if we're done processing
+       * events.
+       */
+      lock_state.unlock();
+      // Assert: event state is `Quiet`
+      lock_queue.lock();
+      // Assert: event state is `Receive`
+      lock_state.lock();
+      // Assert: event state is `Hand-off`
+
```

Comment:
Is there any reason to be concerned about the cost of this locking/unlocking?  (I can't think of any alternative off the top of my head.)